consumer/retry_counter_test.go

```diff
@@ -0,0 +1,97 @@
+package consumer
+
+import (
+	"context"
+	"go.uber.org/goleak"
+	"testing"
+	"time"
+)
+
+func TestRetryCounter(main *testing.T) {
+	main.Run("Is initially zero", func(t *testing.T) {
+		defer goleak.VerifyNone(t)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		rc := newRetryCounter(ctx, make(chan State))
+
+		assertRetryCount(t, rc, 0)
+	})
+
+	main.Run("Is one when unready once", func(t *testing.T) {
+		defer goleak.VerifyNone(t)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		ch := make(chan State)
+		rc := newRetryCounter(ctx, ch)
+
+		ch <- State{
+			Unready: &Unready{},
+		}
+
+		assertRetryCount(t, rc, 1)
+	})
+
+	main.Run("Is two when unready twice", func(t *testing.T) {
+		defer goleak.VerifyNone(t)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		ch := make(chan State)
+		rc := newRetryCounter(ctx, ch)
+
+		ch <- State{
+			Unready: &Unready{},
+		}
+		ch <- State{
+			Unready: &Unready{},
+		}
+
+		assertRetryCount(t, rc, 2)
+	})
+
+	main.Run("Is zero when incremented and reset", func(t *testing.T) {
+		defer goleak.VerifyNone(t)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		ch := make(chan State)
+		rc := newRetryCounter(ctx, ch)
+
+		ch <- State{
+			Unready: &Unready{},
+		}
+		ch <- State{
+			Ready: &Ready{},
+		}
+
+		assertRetryCount(t, rc, 0)
+	})
+}
+
+func assertRetryCount(t *testing.T, rc *retryCounter, expectedRetryCount int) {
```

Comment:
Is this function due to the indeterminiatic state of the read, ie. not knowing when it has read the state change?

It effectively becomes a bust loop while waiting due the `for` and `select default` case.

But let's just push this upstream and request some feedback on this section. 👍

Reply:
I agree. I will close this PR and we will continue on the upstream PR instead.

Reply:
"Is this function due to the indeterminiatic state of the read, ie. not knowing when it has read the state change?" Yes, I could find no way to avoid the loop